_gsocproposals/2020/proposal_ROOTIO.md

```diff
@@ -0,0 +1,49 @@
+The idea of the project is that different filter or transformation schemes that could be applied to ROOT data, which are stored in "baskets". 
+
+In ROOT, the serialization of variable-sized data (containing C-style arrays) produces two internal arrays: one contains the branch data for each event while the other contains the byte offset of each event in the branch data. LZ4 compression algorithm achieves its performance by looking for byte-aligned patterns (as opposed to ZLIB compresison algorithm, which works on individual bits) and lacks the Huffman encoding pass, this results in the offset array sequence being effectively incompressible using LZ4.
+
+Some time ago, Bitshuffle [3] was demonstrated as a possible pre-conditioner for ROOT data with LZ4 for lossless compression [4]. To improve the performance of LZ4 in this case, we investigated the combination of LZ4 with various “pre-conditioners”. Pre-conditioners transform the sequence of input bytes according to a simple, deterministic algorithm prior to applying the compression algorithm. The two algorithms investigated, inspired by the Blosc library [2], are Shuffle and BitShuffle [3]. Both preconditioners rearrange the input array’s bytes by reading through the array using fixed strides. The resulting output of the preconditioner often contains long runs of repeated bytes, improving the compression ratio for LZ4. One of the issues exposed was that it is difficult for ROOT to compress its buffers now due to its 9-byte header.
+
+The goal of the project is to validate the possibility to use pre-conditioners in the ROOT compression layer used to compress both ROOT file formats (TTree and RNtuple) for fast ROOT compression algorithms: LZ4, ZSTD [4].
+
+The second part of the project could be to understand how we can extend feature in the context of both lossy and lossless compression algorithms.
+
+Extra task (if there will be a free time) that could be investigated is new BYTE_STREAM_SPLIT encoding [6] improves compression ratio and compression speed for certain types of floating-point data where the upper-most bytes of a values do not change much. The exisiting compressors and encodings in ROOT do not perform well for such data due to noise in the mantissa bytes. The new encoding improves results by extracting the well compressible bytes into separate byte streams which can be afterwards compressed by a compressor like ZSTD. [5]
+
+## Task ideas
+ * Integrate pre-conditioners in ROOT compression layer (based on our preliminary tests).
+ * Test pre-conditioners for existing fast ROOT compression algorithms.
+ * Validate results for the both ROOT file formats: ROOT TTree and ROOT RNTuple.
+
+## Expected results
+The deliverable of the project should be a detailed report and presentation on performance comparison of the pre-conditioners and compression algorithms used in ROOT. ROOT’s compression library and build system should be extended to support pre-conditioners for compression algorithms. 
```

Comment:
What about a demonstrator showing compression improvement compared to the existing implementation?

Reply:
Done!